### PR TITLE
Bypass v13

### PR DESCRIPTION
**Pull request by regit**
Update of #2282 introducing the local bypass feature.

Local bypass seems to be really interesting as show the result of tests with iperf on loopback with a MTU of 1500:
* standard IPS mode: 669Mbps
* IPS with local bypass: 899Mbps
* IPS with NFQ bypass: 39 Gbps

PR builds:
- PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/201
- PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/197

### CODE REVIEW

File: src/flow-manager.c [hunk 1]
```diff
@@ -67,6 +67,8 @@
 
 #include "output-flow.h"
 
+#define FLOW_TIMEOUT    5
```
Comment (inliniac) [inline src/flow-manager.c]: Needs a more descriptive name.

File: src/flow-worker.c [hunk 2]
```diff
@@ -66,6 +68,13 @@ typedef struct FlowWorkerThreadData_ {
  */
 static inline void FlowUpdate(Packet *p)
 {
+
+    int state = SC_ATOMIC_GET(p->flow->flow_state);
```
Comment (inliniac) [inline src/flow-worker.c]: I would prefer this to move into FlowHandlePacketUpdate, as that is more 'core flow engine'

File: src/flow-worker.c [hunk 3]
```diff
@@ -66,17 +66,15 @@ typedef struct FlowWorkerThreadData_ {
  *
  *  Handle flow creation/lookup
  */
-static inline void FlowUpdate(Packet *p)
-{
-
-    int state = SC_ATOMIC_GET(p->flow->flow_state);
+static inline void FlowUpdate(Packet *p, int state)
```
Comment (inliniac) [inline src/flow-worker.c]: new line issue

File: src/flow-worker.c [hunk 4]
```diff
@@ -170,7 +168,16 @@ TmEcode FlowWorker(ThreadVars *tv, Packet *p, void *data, PacketQueue *preq, Pac
         FlowHandlePacket(tv, fw->dtv, p);
         if (likely(p->flow != NULL)) {
             DEBUG_ASSERT_FLOW_LOCKED(p->flow);
-            FlowUpdate(p);
+            int state = SC_ATOMIC_GET(p->flow->flow_state);
+            FlowUpdate(p, state);
+            switch (state) {
```
Comment (inliniac) [inline src/flow-worker.c]: move this switch into FlowUpdate and make it return a value that we can check for discarding?

File: suricata.yaml.in [hunk 5]
```diff
@@ -1069,26 +1069,34 @@ flow-timeouts:
     new: 30
     established: 300
     closed: 0
+    bypassed: 100
```
Comment (inliniac) [inline suricata.yaml.in]: explanation in yaml missing

File: src/decode.h [hunk 6]
```diff
@@ -447,6 +447,8 @@ typedef struct Packet_
 
     /** The release function for packet structure and data */
     void (*ReleasePacket)(struct Packet_ *);
+    /** The function triggering bypass the flow in the capture method */
```
Comment (inliniac) [inline src/decode.h]: what are the expected retvals?

File: src/source-nfq.c [hunk 7]
```diff
@@ -492,6 +501,21 @@ static void NFQReleasePacket(Packet *p)
     PacketFreeOrRelease(p);
 }
 
+static int NFQBypassCallback(Packet *p)
+{
+    if (IS_TUNNEL_PKT(p)) {
```
Comment (inliniac) [inline src/source-nfq.c]: Needs locking. See TmqhOutputPacketpool

File: suricata.yaml.in [hunk 8]
```diff
@@ -1421,6 +1430,8 @@ nfq:
 #  mode: accept
 #  repeat-mark: 1
 #  repeat-mask: 1
+#  bypass-mark: 1
```
Comment (inliniac) [inline suricata.yaml.in]: yaml doc missing

File: src/decode.c [hunk 9]
```diff
@@ -382,6 +382,24 @@ void PacketDefragPktSetupParent(Packet *parent)
...
+    /* Don't try to bypass if flow is already out or
+     * if we have failed to do it once */
+    int state = SC_ATOMIC_GET(p->flow->flow_state);
+    if ((state == FLOW_STATE_LOCAL_BYPASSED) ||
+           (state == FLOW_STATE_CAPTURE_BYPASSED)) {
+        return;
+    }
+
+    if (p->BypassPacketsFlow && p->BypassPacketsFlow(p)) {
+        /* only set bypassed state if succesful */
```
Comment (inliniac) [inline src/decode.c]: comment not very accurate anymore.

**Comment by inliniac** [comment]
Needs some work.

File: src/decode.c (hunk 9 above)
Comment (regit) [inline src/decode.c]: fixed.

File: suricata.yaml.in (hunk 5 above)
Comment (regit) [inline suricata.yaml.in]: fixed.

File: suricata.yaml.in (hunk 8 above)
Comment (regit) [inline suricata.yaml.in]: fixed.

File: src/decode.h (hunk 6 above)
Comment (regit) [inline src/decode.h]: added in comment.

File: src/flow-worker.c (hunk 3 above)
Comment (regit) [inline src/flow-worker.c]: fixed

**Comment by inliniac** [comment]
On 27-09-16 13:10, Eric Leblond wrote:
> *@regit* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In src/flow-manager.c <https://github.com/inliniac/suricata/pull/2293>:
> 
>> @@ -67,6 +67,8 @@
>  
>  #include "output-flow.h"
>  
> +#define FLOW_TIMEOUT    5
> 
> it is renamed to FLOW_BYPASSED_TIMEOUT in a subsequent commit. Is it
> enough ?

In small sets like this I prefer things to be changed in a single step
where possible.

File: src/flow-manager.c (hunk 1 above)
Comment (regit) [inline src/flow-manager.c]: fixed.

File: src/flow-worker.c (hunk 2 above)
Comment (regit) [inline src/flow-worker.c]: ok.

File: src/flow-worker.c (hunk 4 above)
Comment (regit) [inline src/flow-worker.c]: ok

File: src/source-nfq.c (hunk 7 above)
Comment (regit) [inline src/source-nfq.c]: Fixed.

**Pull request closed without merging**